File: ES_Toolkit_R/R/StationFinder.R
#' Find stations near a park
#' 
#' Uses NOAA Regional Climate Centers' Applied Climate Information System (ACIS) web services \url{http://www.rcc-acis.org/docs_webservices.html}. ACIS overview: \url{http://www.rcc-acis.org/index.html}  ACIS journal reference: \url{https://doi.org/10.1175/BAMS-D-13-00032.1} 
#'
#'
#' Takes one park organizational code and one or more climate parameters, determines the stations near the specified park using a bounding box from the IRMA Unit Service (\url{http://irmaservices.nps.gov/v2/rest/unit/CODE/geography?detail=envelope&dataformat=wkt&format=json}) or from a supplied custom bounding box parameter. 
#' If distance parameter is specified, bounding box will be buffered by that distance. If no distance is provided, park bounding box is used. 
#' Station location must intersect the bounding box (either unbuffered or buffered).
#' Returns station information as a data frame with the following items: name, longitude, latitude, station IDs (sids), state code, elevation (feet), and unique station ID
# @param sourceURL sourceURL for ACIS data services
#' @param unitCode One NPS unit code as a string
#' @param distance (optional) Distance (in kilometers) to buffer park bounding box
#' @param climateParameters A list of one or more climate parameters (e.g. pcpn, mint, maxt, avgt, obst, snow, snwd). If not specified, defaults to all parameters except degree days. See Table 3 on ACIS Web Services page: \url{http://www.rcc-acis.org/docs_webservices.html}
#' @param filePathAndName (optional) File path and name including extension for output CSV file
#' @param customBBox (optional) String containing bounding box geographic coordinates (longitude,latitude) using the WGS84 datum in the following format: Lower Left Longitude, Lower Left Latitude, Upper Right Longitude, Upper Right Latitude. Note: longitude is negative in the western hemisphere. Example: "-114.291153779, 35.5612153111, -111.252315168, 37.0351548001"
#' @return A data frame containing station information for stations near the specified park. See User Guide for more details:  \url{}
#' @examples \dontrun{
#' Find stations collecting average temperature within 10km of Marsh-Billings NHP:
#' 
#' findStation(unitCode = "MABI", distance=10, climateParameters=list('avgt'))
#' 
#' Find stations collecting all climate parameters except degree days within 15km of Marsh-Billings NHP:
#' 
#' findStation(unitCode = "MABI", distance=10)
#' 
#' Find stations collecting precipitation or average temperature within 10km of Agate Fossil Beds and save to a CSV file:
#' 
#' findStation(unitCode = "AGFO", distance=10, climateParameters=list('pcpn'), filePathAndName = "agfo_stations.csv")
#' 
#' Find stations within 30km of Rocky Mountain NP collecting maxt and mint:
#' 
#' findStation(unitCode = "ROMO", distance=30, climateParameters=list('pcpn'), filePathAndName = "Test01_R.csv")
#' 
#'Find stations within 50km of Alamosa NWR that collect precipitation data 
#' }
#' @export 
#' 

# TODO: iterate unitCode list; add either/or capability for park code/bbox

findStation <- function (unitCode, distance=NULL, climateParameters=NULL, filePathAndName=NULL, customBBox=NULL) {
  # URLs and request parameters
  
  # NPS Park bounding boxes
  if (is.null(distance)) {
    bboxExpand  = 0.0
  }
  else if (distance == 0) {
    bboxExpand = 0.0
  }
  else {
    bboxExpand = distance*0.011  # convert km to decimal degrees
  }
  
  # ACIS data services
  baseURL <- "http://data.rcc-acis.org/"
  webServiceSource <- "StnMeta"
  lookups <- 
    fromJSON(system.file("ACISLookups.json", package = "EnvironmentalSettingToolkit"), flatten = TRUE) # assumes placement in package inst subfolder
  
  stationMetadata = c('uid', 'name', 'state', 'll', 'elev', 'valid_daterange', 'sids')
  #stationMetadata <-c('uid', 'name', 'state', 'll', 'elev', 'valid_daterange', 'sids')
  # If climateParameters is NULL, default to all parameters except degree days.
  #parameters <- list('pcpn', 'avgt', 'obst', 'mint', 'maxt', 'snwd', 'snow') 
  encode <- c("json")
  config <- add_headers(Accept = "'Accept':'application/json'")
  
  stationURL <- gsub(" ","",paste(baseURL,webServiceSource))
  
  #Example URLS
  # http://data.rcc-acis.org/StnMeta?bbox=-104.895308730118,%2041.8657116369158,%20-104.197521654032,%2042.5410939149279&meta=uid,%20name,%20state,%20ll,%20elev,%20valid_daterange,sids
  # http://data.rcc-acis.org/StnMeta?bbox=-104.895308730118,%2041.8657116369158,%20-104.197521654032,%2042.5410939149279
  
  # Get bounding box for park(s)
  if (is.null(customBBox)) {
    bbox <- getBBox(unitCode, bboxExpand)
  }
  else {
    bbox <- getBBox(unitCode, bboxExpand, customBBox)
  }
  body  <- list(bbox = bbox)

  # Format GET URL for use in jsonlite request
  if (is.null(climateParameters)) {
    climateParameters0 <- lookups$element$code
    # Remove degree days (v1.5); super cheesy... fix at some point
    climateParameters <- climateParameters0[1:7]
    #climateParameters = parameters
  }
  stationList = NULL
  for (p in 1:length(climateParameters)) {
    stationRequest <- gsub(" ", "%20", paste(paste(paste(stationURL, paste(climateParameters[p], collapse = ","), sep="?elems="), body, sep="&bbox="), paste(stationMetadata, collapse=","), sep="&meta="))
    #stationRequest <- gsub(" ", "%20", paste(paste(paste(stationURL, paste(climateParameters, collapse = ","), sep="?elems="), body, sep="&bbox="), paste(stationMetadata, collapse=","), sep="&meta="))
    #stationRequest <- gsub(" ", "%20", paste(paste(stationURL, paste(climateParameters, collapse = ","), sep="?elems="), body, sep="&bbox="))
    #stationRequest <- gsub(" ", "%20", paste(paste(paste(stationURL, paste(climateParameters, collapse = ","), sep="?elems="), body, sep="&bbox=")),paste(stationMetadata, collapse = ","), sep="&meta=")
    
    # Use bounding box to request station list (jsonlite)
    stationListInit <- fromJSON(stationRequest) 
    if (length(stationListInit$meta) > 0) {
      uid <- setNames(as.data.frame(as.numeric(stationListInit$meta$uid)), "uid")
      longitude <- setNames(as.data.frame(as.numeric(as.matrix(lapply(stationListInit$meta$ll, function(x) unlist(as.numeric(x[1])))))),"longitude")
      latitude <- setNames(as.data.frame(as.numeric(as.matrix(lapply(stationListInit$meta$ll, function(x) unlist(as.numeric(x[2])))))),"latitude")
      # Check for presence of all SID values (use max of 3 per record even if station has > 3)
      # Suppress warnings from getStationSubtype(): raised due to conversion necessary because data.frame vector access does not recognize column name
      sid1 = c()
      sid2 = c()
      sid3 = c()
      sid1_type = c(as.character(NA))
      sid2_type = c(as.character(NA))
      sid3_type = c(as.character(NA))
      minDate = c(as.Date(NA))
      maxDate = c(as.Date(NA))
      hcn = c()
      for (i in 1:length(stationListInit$meta$sids)) {
        if (length(unlist(stationListInit$meta$sids[i])) >= 3) {
          sid1[i] <- as.character(as.vector(lapply(stationListInit$meta$sids[i], function(x) unlist(x[1]))))
          sid1_type[i] <-  suppressWarnings(getStationSubtype(unlist(strsplit(sid1[i], " "))[2], substr(sid1[i],1,3)))
          
          sid2[i] <- as.character(as.vector(lapply(stationListInit$meta$sids[i], function(x) unlist(x[2]))))
          sid2_type[i] <-  suppressWarnings(getStationSubtype(unlist(strsplit(sid2[i], " "))[2], substr(sid2[i],1,3)))
          sid3[i] <- as.character(as.vector(lapply(stationListInit$meta$sids[i], function(x) unlist(x[3]))))
          sid3_type[i] <-  suppressWarnings(getStationSubtype(unlist(strsplit(sid3[i], " "))[2], substr(sid3[i],1,3)))
        }
        else if (identical(length(unlist(stationListInit$meta$sids[i])), as.integer(c(2)))) {
          sid1[i] <- as.character(as.vector(lapply(stationListInit$meta$sids[i], function(x) unlist(x[1]))))
          sid1_type[i] <-  suppressWarnings(getStationSubtype(unlist(strsplit(sid1[i], " "))[2], substr(sid1[i],1,3)))
          sid2[i] <- as.character(as.vector(lapply(stationListInit$meta$sids[i], function(x) unlist(x[2]))))
          sid2_type[i] <-  suppressWarnings(getStationSubtype(unlist(strsplit(sid2[i], " "))[2], substr(sid2[i],1,3)))
          sid3[i] <- as.character(NA)
          sid3_type[i] <-  as.character(NA)
        }
        else {
          sid1[i] <- as.character(as.vector(lapply(stationListInit$meta$sids[i], function(x) unlist(x[1]))))
          sid1_type[i] <-  suppressWarnings(getStationSubtype(unlist(strsplit(sid1[i], " "))[2], substr(sid1[i],1,3)))
          sid2[i] <- as.character(NA)
          sid2_type[i] <-  as.character(NA)
          sid3[i] <- as.character(NA)
          sid3_type[i] <-  as.character(NA)
        }
      }
      #sid1 <- setNames(sid1,"sid1")
      sid1 <- setNames(as.data.frame(sid1),"sid1")
      sid2 <- setNames(as.data.frame(sid2),"sid2")
      sid3 <- setNames(as.data.frame(sid3),"sid3")
      sid1_type <- setNames(as.data.frame(sid1_type),"sid1_type")
      sid2_type <- setNames(as.data.frame(sid2_type),"sid2_type")
      sid3_type <- setNames(as.data.frame(sid3_type),"sid3_type")
      i <- NULL
      for (i in 1:length(stationListInit$meta$sids)) {
        if (!is.null(unlist(stationListInit$meta$valid_daterange[i]))) {
          minDate[i] <- as.Date(range(unlist(stationListInit$meta$valid_daterange[i]))[1], "%Y-%m-%d")
          maxDate[i] <- as.Date(range(unlist(stationListInit$meta$valid_daterange[i]))[2], "%Y-%m-%d")
        }
        else {
          minDate[i] <- NA
          maxDate[i] <- NA
        }
      }
      minDate <-  setNames(as.data.frame(minDate), "minDate")
      maxDate <-  setNames(as.data.frame(maxDate), "maxDate")
      # Detect U.S. Historical Climate Network stations using the first station identifier
      hcn0 <- getUSHCN(sid1)
      hcn <- setNames(as.data.frame(hcn0), "isHCNStation")
      browser()
      # Force elevation to be numeric with precision of 1
      options(digits = 1)
      if (!is.null(stationListInit$meta$elev)) {
        elev <- as.numeric(stationListInit$meta$elev)
      }
      else {
        elev <- as.numeric(NA)
      }
      options(digits = 7)
      if (is.null(stationList)) {
        stationList <- cbind( uid, name=stationListInit$meta$name, longitude, latitude, sid1, sid1_type, sid2, sid2_type, sid3, sid3_type, state=stationListInit$meta$state, elev=elev, isHCNStation=hcn, minDate, maxDate)
        #stationList <- cbind( uid, name=stationListInit$meta[,1], longitude, latitude, sid1, sid1_type, sid2, sid2_type, sid3, sid3_type, state=stationListInit$meta[,4], elev=stationListInit$meta[,5], minDate, maxDate)
        stationList$climateParameter <- unlist(climateParameters[p])
        stationList$unitCode <- unitCode[1]
      }
      else {
        stationListTemp <- cbind( uid, name=stationListInit$meta$name, longitude, latitude, sid1, sid1_type, sid2, sid2_type, sid3, sid3_type, state=stationListInit$meta$state, elev=elev, isHCNStation=hcn, minDate, maxDate)
        #stationListTemp <- cbind( uid, name=stationListInit$meta[,1], longitude, latitude, sid1, sid1_type, sid2, sid2_type, sid3, sid3_type, state=stationListInit$meta[,4], elev=stationListInit$meta[,5], minDate, maxDate)
        stationListTemp$climateParameter <- unlist(climateParameters[p])
        stationListTemp$unitCode <- unitCode[1]
        stationList <- rbind(stationList, stationListTemp)
      }
      # Convert factors to character vectors
      fc  <- sapply(stationList, is.factor)
      lc <- sapply(stationList, is.logical)
      stationList[, fc]  <- sapply(stationList[, fc], as.character)
      stationList[, lc]  <- sapply(stationList[, lc], as.character)
    }
    else {
      stationList <- NA # per John Paul's request - Issue #49
      #stationList <- cat("No stations for ", unitCode, "using distance ", distance) 
    }
  }
  # Output file
  if (!is.null(filePathAndName)) {
    if (file.exists(filePathAndName)) {
      write.table(stationList, file=filePathAndName, append = TRUE, sep=",", row.names=FALSE, col.names=FALSE, qmethod="double")
    }
    else {
      write.table(stationList, file=filePathAndName, sep=",", row.names=FALSE, qmethod="double")
    }
  }
  else {
    return (stationList)
  }
  return (stationList)
}

File: ES_Toolkit_py/StationFinder.py
import json
import urllib2

from StationDict import StationDict
from WxOb import WxOb
from ACIS import ACIS

class StationFinder(ACIS):
    '''
    INFO
    -------
    Object to find weather stations using ACIS Web Services.

    '''
    interval = None     #Not applicable when finding stations
    add = None          #Not applicable when finding stations
    duration = None     #Not applicable when finding stations
    reduceCodes = []    #Not applicable when finding stations

    def __init__(self, *args, **kwargs):
        super(StationFinder, self).__init__(*args, **kwargs)
        self.webServiceSource = 'StnMeta'

    def _formatElems(self):
        '''
        Unfortunately, the formatting of elements is different when requesting
        a list of stations. Thus, there is the need to override the standard ACIS
        method.
        '''
        self.elems = self.climateParameters

    def findStation(self, unitCode = None, distance = 0,
        climateParameters = None, sdate = None, edate = None
        ,filePathAndName = None, customBBox = None):
        '''
        INFO
        ----
        Standard method to find all stations and associated metadata
        based on zero or more criteria.


        ARGUMENTS
        ---------
        unitCode           4-Letter park code (searches for station within buffer)

        distance            buffer distance around the provided unitCode (if provided).
                            Default is 0 km.

        climateParameters    Parameter code for climate/weather element (e.g., mint, avgt, pcpn)

        sdate               (Optional) - Start date as yyyy-mm-dd or yyyymmdd.
                             If not provided, the default is the period of record.

        edate               (Optional) - End date as yyyy-mm-dd or yyyymmdd.
                             If not provided, the default is the period of record

        filePathAndName    If provided, a csv text file is saved to specific location.

        customBBox (optional) String containing bounding box geographic coordinates (longitude,latitude) using the WGS84 datum in the following format: Lower Left Longitude, Lower Left Latitude, Upper Right Longitude, Upper Right Latitude. Note: longitude is negative in the western hemisphere. Example: "-114.291153779, 35.5612153111, -111.252315168, 37.0351548001".

        RETURNS
        -------
        An object of station metadata (See StationDict.py)
        '''

        #ACIS handles start and end dates differently when searching for stations
        if not sdate:
            sdate = 'NA'
        if not edate:
            edate = 'NA'

        metadata = ('uid', 'name', 'state', 'll', 'elev', 'valid_daterange', 'sids')

        kwargs = self._formatArguments(unitCode = unitCode, distance = distance
            , climateParameters = climateParameters, reduceCodes = self.reduceCodes,
            sdate = sdate, edate = edate, meta = metadata, customBBox = customBBox )

        results =  self._call_ACIS(kwargs = kwargs)
        self._checkResponseForErrors(results)

        si = StationDict(observationClass = WxOb, queryParameters = self._input_dict, climateParameters = self.climateParameters)
        # Convert pandas.numpy array to list (plain text)
        hcnLookup = self._hcn_lookups.coopid.values.tolist()
        #if hcnLookup[5:10].strip() in station['sids']:
        for station in results['meta']:
            station['unitCode'] = unitCode #added to associate each station with a unit code
            #if station['sids'].split(" ") in self._hcn_lookups.coopid.values:
            # sids element is a list of unicode strings
            if (x[5:10].encode('latin1') for x in station['sids'] if x in y[5:10] for y in hcnLookup):
                station['isHCNStation'] = 'Y'
            else:
                station['isHCNStation'] = 'N'
            si._addStation(stationID = station['uid'], stationMeta = station)
        if filePathAndName:
                si.exportMeta(filePathAndName)
        return si

if __name__ == '__main__':
    sf = StationFinder()

    wxStations = sf.findStation(unitCode = 'NOCA',
        sdate = '1940-01-01', edate = '1940-01-01')
    print wxStations.queryParameters
    print wxStations
    print sf.supportedParameters
    print wxStations[26202].validDateRange
    print wxStations[26202].validDateRange.keys()

    wxStations = sf.findStation(unitCode = 'ROMO', climateParameters = 'mint, maxt,pcpn')
    print wxStations
    print wxStations[48106].validDateRange
    print wxStations[48106].validDateRange['pcpn']

    wxStations = sf.findStation(unitCode = 'ROMO', climateParameters = ['mint', 'maxt','pcpn'])
    print wxStations[48106].validDateRange

    stationList = sf.findStation(unitCode = 'AGFO', distance = 10, climateParameters = 'pcpn'
        ,sdate = '2015-08-01', edate = '2015-08-04')
    print stationList

    wxStations = sf.findStation(unitCode = 'NOCA',
         sdate = '2014-01-01', edate = '2016-01-01')
    print wxStations


